### PR TITLE
Pipeline redesign for object_detection_demo_ssd_async (C++)

### PR DESCRIPTION
**Pull request by AnthonyQuantum**

### CODE REVIEW

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 1]
```diff
@@ -229,100 +281,93 @@ int main(int argc, char *argv[]) {
...
 
-        // --------------------------- 6. Do inference ---------------------------------------------------------
-        slog::info << "Start inference " << slog::endl;
-
-        bool isLastFrame = false;
-        bool isAsyncMode = false;  // execution is always started using SYNC mode
-        bool isModeChanged = false;  // set to TRUE when execution mode is changed (SYNC<->ASYNC)
+        // --------------------------- 6. Init variables -------------------------------------------------------
+        typedef std::chrono::duration<double, std::chrono::milliseconds::period> ms;
+        typedef std::chrono::duration<double, std::chrono::seconds::period> sec;
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:297]: They represent types, so the names should be Ms and Sec.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 2]
```diff
@@ -229,100 +281,93 @@ int main(int argc, char *argv[]) {
...
+        struct RequestResult {
+            cv::Mat frame;
+            const float* output;
+            std::chrono::high_resolution_clock::time_point startTime;
+            bool isSameMode;
+        };
+
+        struct ModeInfo {
+           int framesCount = 0;
+           double latencySum = 0;
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:308]: You should use there the original type `std::chrono::steady_clock::duration`. Extract printable value only when you need to report it. With that approach you won't accumulate errors produced by double representation.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 3]
```diff
@@ -229,100 +281,93 @@ int main(int argc, char *argv[]) {
...
+                     "press ESC or 'q' key" << std::endl;
+        std::cout << "To switch between min_latency/user_specified modes, press TAB key in the output window" 
+                  << std::endl;
+
+        while ((cap.isOpened()
+                || !completedRequestResults.empty()
+                || ((isUserSpecifiedMode && emptyRequests.size() < FLAGS_nireq)
+                    || (!isUserSpecifiedMode && emptyRequests.size() < 1)))
+               && callbackException == nullptr) {
+            if (callbackException) std::rethrow_exception(callbackException);
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:350]: You should protect some of these vars because they are modified by another thread. I suggest moving the condition under the scope of the following `lock_guard`.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 4]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+
+                std::ostringstream out;
+                out << "FPS: " << std::fixed << std::setprecision(1) << modeInfo[isUserSpecifiedMode].framesCount / 
+                    std::chrono::duration_cast<sec>(std::chrono::high_resolution_clock::now() -
+                                                    modeInfo[isUserSpecifiedMode].lastStartTime).count();
+                cv::putText(requestResult.frame, out.str(), cv::Point2f(10, 25), cv::FONT_HERSHEY_TRIPLEX, 0.6, 
+                            cv::Scalar(0, 0, 255), 1);
+                out.str("");
+                modeInfo[isUserSpecifiedMode].latencySum += std::chrono::duration_cast<sec>(
+                    std::chrono::high_resolution_clock::now() - requestResult.startTime).count();
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:414]: You should have a variable storing the result of `now()`, so you don't need to call it twice and you won't get different values.

Actually the same idea should have been applied to `object_detection_demo_yolov3_async.py`, but I didn't notice that back than.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 5]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                out << "Latency: " << std::fixed << std::setprecision(1) << (modeInfo[isUserSpecifiedMode].latencySum /
+                    modeInfo[isUserSpecifiedMode].framesCount) * 1e3 << " ms";
+                cv::putText(requestResult.frame, out.str(), cv::Point2f(10, 50), cv::FONT_HERSHEY_TRIPLEX, 0.6,
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:417]: I think you should add `put_highlighted_text()` as for `object_detection_demo_yolov3_async.py`.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 6]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                    cv::imshow("Detection Results", requestResult.frame);
+                    
+                    const int key = cv::waitKey(1);
+
+                    if (27 == key || 'q' == key || 'Q' == key) {  // Esc
+                        break;
+                    }
+                    else if (9 == key) {  // Tab
+                        bool prevMode = isUserSpecifiedMode;
+                        isUserSpecifiedMode ^= true;
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:430]:
```suggestion
                        isUserSpecifiedMode = !isUserSpecifiedMode;
```

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 7]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                        } else emptyRequests.push(minLatencyInferRequest);
+
+                        modeInfo[prevMode].lastEndTime = std::chrono::high_resolution_clock::now();
+                        modeInfo[isUserSpecifiedMode] = ModeInfo();
+                    } else {
+                        presenter.handleKey(key);
+                    }
+                }
             }
+            else if (!emptyRequests.empty() && cap.isOpened()) {
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:453]:
```suggestion
            } else if (!emptyRequests.empty() && cap.isOpened()) {
```

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 8]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
-            t1 = std::chrono::high_resolution_clock::now();
-            ocv_render_time = std::chrono::duration_cast<ms>(t1 - t0).count();
+                InferRequest::Ptr request;
+                {
+                    std::lock_guard<std::mutex> lock(mutex);
 
-            if (isLastFrame) {
-                break;
-            }
+                    request = emptyRequests.front();
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:472]:
```suggestion
                    request = std::move(emptyRequests.front());
```

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 9]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                }
+                frameToBlob(frame, request, imageInputName);
+                
+                bool frameMode = isUserSpecifiedMode;
+                request->SetCompletionCallback([&mutex,
+                                                &completedRequestResults,
+                                                nextFrameId,
+                                                frame,
+                                                request,
+                                                outputName,
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:483]:
```suggestion
                                                &outputName,
```

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 10]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                                                &emptyRequests,
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame, request->GetBlob(outputName)->buffer().as<float*>(),
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:496]: it should be rmap()

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 11]
```diff
@@ -346,59 +391,188 @@ int main(int argc, char *argv[]) {
...
+                    }
+                    condVar.notify_one();
+                });
 
-            if (isModeChanged) {
-                isModeChanged = false;
+                request->StartAsync();
+                nextFrameId++;
             }
+            else {
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/main.cpp:515]:
```suggestion
            } else {
```

File: demos/object_detection_demo_ssd_async/README.md [hunk 12]
```diff
@@ -3,16 +3,13 @@
 This demo showcases Object Detection with SSD and new Async API.
 Async API usage can improve overall frame-rate of the application, because rather than wait for inference to complete,
 the app can continue doing things on the host, while accelerator is busy.
-Specifically, this demo keeps two parallel infer requests and while the current is processed, the input frame for the next
-is being captured. This essentially hides the latency of capturing, so that the overall framerate is rather
-determined by the `MAXIMUM(detection time, input capturing time)` and not the `SUM(detection time, input capturing time)`.
+Specifically, this demo keeps the number of Infer Requests that you have set using input flag. While some of the Infer Requests are processed, the other ones can be filled with new frame data and asynchronously started or the next output can be taken from the Infer Request and displayed.
```
Comment (Wovchena) [inline demos/object_detection_demo_ssd_async/README.md:6]:
```suggestion
Specifically, this demo keeps the number of Infer Requests that you have set using `nireq` flag. While some of the Infer Requests are processed by IE, the other ones can be filled with new frame data and asynchronously started or the next output can be taken from the Infer Request and displayed.
```

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 10 above)
Comment (Wovchena) [reply demos/object_detection_demo_ssd_async/main.cpp]: There the problem with lifetime of `LockedMemory`. `outputMapped.as<float*>()` is stored in `RequestResult` while `outputMapped` is destroyed.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 13]
```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
...
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::unique_ptr<LockedMemory<const void>>(new
+                                        LockedMemory<const void>(as<MemoryBlob>(request->GetBlob(outputName))->rmap())),
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:509]: Why do you map the blob here? IMO, it would be a lot clearer to do it where the contents of the blob are actually used.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 14]
```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
...
-                isModeChanged = true;
+                request->StartAsync();
+                nextFrameId++;
             } else {
-                presenter.handleKey(key);
+                std::unique_lock<std::mutex> lock(mutex);
+
+                while (callbackException == nullptr && emptyRequests.empty() && completedRequestResults.empty()) {
+                    condVar.wait(lock);
+                }
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:534]: This seems odd. You wait for one of these objects to be updated from a background thread, but then you immediately release the lock and access those objects while the lock is _not_ held - isn't that thread-unsafe?

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 15]
```diff
@@ -235,102 +298,88 @@ int main(int argc, char *argv[]) {
...
+            emptyRequests.assign(userSpecifiedInferRequests.begin(), userSpecifiedInferRequests.end());
+        } else emptyRequests = {minLatencyInferRequest};
+
+        std::unordered_map<int, RequestResult> completedRequestResults;
+        int nextFrameId = 0;
+        int nextFrameIdToShow = 0;
+        std::exception_ptr callbackException = nullptr;
+        std::mutex mutex;
+        std::condition_variable condVar;
+        std::map<bool, ModeInfo> modeInfo = {{true, ModeInfo()}, {false, ModeInfo()}};
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:342]: The usage of `bool` for the execution mode makes the code hard to understand, because it's not obvious which is `true` and which is `false`. I would suggest introducing an `enum class ExecutionMode` instead.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 16]
```diff
@@ -354,58 +403,194 @@ int main(int argc, char *argv[]) {
...
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::unique_ptr<LockedMemory<const void>>(new
+                                        LockedMemory<const void>(as<MemoryBlob>(request->GetBlob(outputName))->rmap())),
+                                    startTime,
+                                    frameMode == isUserSpecifiedMode
+                                }));
+                            
+                            if (isUserSpecifiedMode == frameMode) {
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:514]: I don't quite understand why this condition is here. Why wouldn't we want to return `request` to `emptyRequests` if the execution mode changed?

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 17]
```diff
@@ -113,14 +122,64 @@ int main(int argc, char *argv[]) {
...
+                }
+
+                // for CPU execution, more throughput-oriented execution via streams
+                userSpecifiedConfig.insert({ CONFIG_KEY(CPU_THROUGHPUT_STREAMS),
+                                (deviceNstreams.count(device) > 0 ? std::to_string(deviceNstreams.at(device))
+                                                                  : CONFIG_VALUE(CPU_THROUGHPUT_AUTO)) });
+
+                minLatencyConfig.insert({ CONFIG_KEY(CPU_THROUGHPUT_STREAMS), "1" });
+
+                deviceNstreams[device] = std::stoi(
```
[the remark this answers is not in the document]
Comment (AnthonyQuantum) [reply demos/object_detection_demo_ssd_async/main.cpp]: Removed

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 13 above)
Comment (AnthonyQuantum) [reply demos/object_detection_demo_ssd_async/main.cpp]: Done

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 14 above)
Comment (AnthonyQuantum) [reply demos/object_detection_demo_ssd_async/main.cpp]: I've replaced `while` with `if` because `while` is not actually useful here, as experiments have shown.

security_barrier_camera_demo and classification_demo might need to be fixed too, because it looks like they have similar problem:
```
std::unique_lock<std::mutex> lk(tasksMutex);
while (running && tasks.empty()) {
        tasksCondVar.wait(lk);
}
```

```
{
        std::unique_lock<std::mutex> lock(mutex);

        while (!irCallbackException
                  && emptyInferRequests.empty() && completedInferRequests.empty()) {
                condVar.wait(lock);
        }
}
```

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 16 above)
Comment (AnthonyQuantum) [reply demos/object_detection_demo_ssd_async/main.cpp]: It's because the `emptyRequests` container is only for requests from current mode. I've changed the name to make it clearer.

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 14 above)
Comment (IRDonch) [reply demos/object_detection_demo_ssd_async/main.cpp]: > I've replaced while with if because while is not actually useful here, as experiments have shown.

No, `while` is necessary because of spurious wakeups.

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 16 above)
Comment (IRDonch) [reply demos/object_detection_demo_ssd_async/main.cpp]: > It's because the emptyRequests container is only for requests from current mode.

Okay, but why, though? When you switch modes, you wait for all pending requests to complete and then reinitialize `emptyRequests`, so does it even matter which requests it contains at that point?

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 18]
```diff
@@ -354,58 +401,197 @@ int main(int argc, char *argv[]) {
...
+                                                &callbackException,
+                                                &condVar] {
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    request->GetBlob(outputName),
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:509]: Given that `request` is later moved to `currentModeEmptyRequests`, isn't it possible that it'll get reused before the result is processed?

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 19]
```diff
@@ -235,102 +292,92 @@ int main(int argc, char *argv[]) {
...
+           int framesCount = 0;
+           std::chrono::steady_clock::duration latencySum;
+           std::chrono::steady_clock::time_point lastStartTime;
+           std::chrono::steady_clock::time_point lastEndTime;
+        };
+
+        auto totalT0 = std::chrono::steady_clock::now();
+
+        // execution always starts in USER_SPECIFIED mode
+        enum class ExecutionMode {userSpecified, minLatency};
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:324]:
```suggestion
        enum class ExecutionMode {USER_SPECIFIED, MIN_LATENCY};
```

Not only is this the convention for enum constants, but you even wrote `USER_SPECIFIED` yourself above. :)

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 14 above)
Comment (IRDonch) [reply demos/object_detection_demo_ssd_async/main.cpp]: You haven't really fixed it yet. I can see that both `callbackException` and `emptyRequests` are accessed in the main loop while the mutex is not held.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 20]
```diff
@@ -354,58 +401,197 @@ int main(int argc, char *argv[]) {
...
+                    {
+                        std::lock_guard<std::mutex> callbackLock(mutex);
+                    
+                        try {
+                            completedRequestResults.insert(
+                                std::pair<int, RequestResult>(nextFrameId, RequestResult{
+                                    frame,
+                                    std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outputName))),
+                                    startTime,
+                                    frameMode == currentMode
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:510]: It would be simpler to just save `frameMode` and compare it to `currentMode` in the main loop (when you're processing the completed request). This way, you won't have to capture `currentMode` in the callback, nor access it from multiple threads without synchronization.

File: demos/object_detection_demo_ssd_async/main.cpp [hunk 21]
```diff
@@ -354,58 +401,197 @@ int main(int argc, char *argv[]) {
...
-                break;
-            if (9 == key) {  // Tab
-                isAsyncMode ^= true;
-                isModeChanged = true;
+                request->StartAsync();
+                nextFrameId++;
             } else {
-                presenter.handleKey(key);
+                while (true) {
+                    std::unique_lock<std::mutex> lock(mutex);
```
Comment (IRDonch) [inline demos/object_detection_demo_ssd_async/main.cpp:528]: Why did you move the lock inside the loop? There's no reason to repeatedly relock the mutex, the original loop structure was fine.

File: demos/object_detection_demo_ssd_async/main.cpp (hunk 18 above)
Comment (AnthonyQuantum) [reply demos/object_detection_demo_ssd_async/main.cpp]: Now it will be reused before the result is processed, but `output` was already copied into the new `RequestResult` object here:
```
completedRequestResults.insert(
                            std::pair<int, RequestResult>(nextFrameId, RequestResult{
                                frame,
                                std::make_shared<TBlob<float>>(*as<TBlob<float>>(request->GetBlob(outputName))),
                                startTime,
                                frameMode
                            }));
```
So, I suppose that everything is safe now.
